### PR TITLE
[DOC] Add in docs about memory debugging [skip ci]

### PR DESCRIPTION
**Pull request by revans2**
This fixes https://github.com/NVIDIA/spark-rapids/issues/9987

any feedback is appreciated.

### CODE REVIEW

**Comment by revans2** [comment]
build

File: docs/dev/README.md [hunk 1]
```diff
@@ -241,6 +242,11 @@ port 5005.
 
 You can also use [Compute Sanitizer](compute_sanitizer.md) to debug CUDA memory errors.
 
+### Memory Debugging
+Java's garbage collector does not play nicely with CUDA memory allocations or with off heap memory.
+There are a number of tools that we have developed that can help do
```
Comment (jbrennan333) [inline docs/dev/README.md:247]:
```suggestion
There are a number of tools that we have developed that can help to
```

File: docs/dev/mem_debug.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+Be aware that there are a lot of things to keep in mind here. Our spill framework often will
+wait for an allocation to fail before it starts to spill, so seeing a `allocate failure` does
+not mean that something terrible happened. `Size` is the size for the allocation at this point.
+Because RMM allows for different layers of code in between the call to RMM and the logging the
+size might be different from what the user actually requested. RMM often will try and "align" the
+allocation. Meaning round it up to a multiple of a set number of bytes. We try to get the logger as
+close to the actual allocator as possible. But just be careful.
+
+Also know that the address here should correspond to the address in the leak debugging if and only
+if it was a DeviceMemoryBuffer that was allocated. In this case, where it is a `cudf::column` the
```
Comment (jbrennan333) [inline docs/dev/mem_debug.md:136]:
```suggestion
if it was a `DeviceMemoryBuffer` that was allocated. In this case, where it is a `cudf::column` the
```

File: docs/dev/mem_debug.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+We don't have a good way to [track spill](https://github.com/NVIDIA/spark-rapids/issues/8752)/log 
+what is or is not spillable. The spill framework typically will use reference counting to know
+when it is the only one holding a reference to memory and then make the data spillable at that
+point.  This means that if we don't get the reference counting right we end up never spilling
+memory, and it may mask the leak for quite a while.
+
+We also don't have a way to 
+[log exactly what was spilled](https://github.com/NVIDIA/spark-rapids/issues/10103)
+and what was read back it. We can probably guess that this is happening from other logs, but it
```
Comment (jbrennan333) [inline docs/dev/mem_debug.md:172]:
```suggestion
and what was read back. We can probably guess that this is happening from other logs, but it
```

**Review by jbrennan333 — DISMISSED** [review]
Looks good.  Just a few nits.

**Comment by revans2** [comment]
@jbrennan333 please take another look

**Comment by revans2** [comment]
build

**Review by jbrennan333 — APPROVED** [review]

**Pull request merged**